core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java

```diff
@@ -285,21 +287,86 @@ public Boolean areColumnsUnique(Intersect rel, RelMetadataQuery mq,
       return mq.areColumnsUnique(left, columns, ignoreNulls);
     }
 
+    int numLeftCols = rel.getLeft().getRowType().getFieldCount();
+
     // Divide up the input column mask into column masks for the left and
     // right sides of the join
     final Pair<ImmutableBitSet, ImmutableBitSet> leftAndRightColumns =
-        splitLeftAndRightColumns(rel.getLeft().getRowType().getFieldCount(),
+        splitLeftAndRightColumns(numLeftCols,
             columns);
-    final ImmutableBitSet leftColumns = leftAndRightColumns.left;
-    final ImmutableBitSet rightColumns = leftAndRightColumns.right;
+    final ImmutableBitSet initLeftColumns = leftAndRightColumns.left;
+    final ImmutableBitSet initRightColumns = leftAndRightColumns.right;
 
     // for FULL OUTER JOIN if columns contain column from both inputs it is not
     // guaranteed that the result will be unique
     if (!ignoreNulls && rel.getJoinType() == JoinRelType.FULL
-        && leftColumns.cardinality() > 0 && rightColumns.cardinality() > 0) {
+        && initLeftColumns.cardinality() > 0 && initRightColumns.cardinality() > 0) {
       return false;
     }
 
+    // If we have an inner join on either side, we may be able to determine
+    // uniqueness from the condition.
+    HashSet<RexCall> equalities = findCommonEqualities(rel.getCondition());
+    // Check each equality. If we are only checking uniqueness with a single
+    // side we can include the other side.
+    BitSet addedLeftCols = new BitSet();
+    BitSet addedRightCols = new BitSet();
+    boolean isOuterLeft = rel.getJoinType().generatesNullsOnRight();
+    boolean isOuterRight = rel.getJoinType().generatesNullsOnLeft();
+    for (RexCall equals: equalities) {
+      Integer arg0 = ((RexInputRef) equals.getOperands().get(0)).getIndex();
+      Integer arg1 = ((RexInputRef) equals.getOperands().get(1)).getIndex();
+      if (arg0 < numLeftCols && arg1 < numLeftCols) {
+        // Both columns are in the left table.
+        if (!isOuterLeft) {
+          // An outer join may not have the columns equal
+          boolean foundArg0 = initLeftColumns.get(arg0);
+          boolean foundArg1 = initLeftColumns.get(arg1);
+          if (foundArg0 && !foundArg1) {
+            addedLeftCols.set(arg1);
+          } else if (!foundArg0 && foundArg1) {
+            addedLeftCols.set(arg0);
```

Comment:
Can't you just do the set for both args? The only purpose of addedLeftCols is to be unioned with the initColumnsSet, so setting an extra bit in addedLeftCols that is already present in the initColumnsSet shouldn't do anything.

Reply:
Good point. I still need to check the condition for knowing if I should keep the test, but I can always set both.